pkg/controller/imagelistpulljob/imagelistpulljob_controller.go

```diff
@@ -0,0 +1,422 @@
+	"k8s.io/kubernetes/pkg/util/slice"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	appsv1alpha1 "github.com/openkruise/kruise/apis/apps/v1alpha1"
+	"github.com/openkruise/kruise/pkg/features"
+	"github.com/openkruise/kruise/pkg/util"
+	utilclient "github.com/openkruise/kruise/pkg/util/client"
+	utildiscovery "github.com/openkruise/kruise/pkg/util/discovery"
+	"github.com/openkruise/kruise/pkg/util/expectations"
+	utilfeature "github.com/openkruise/kruise/pkg/util/feature"
+	"github.com/openkruise/kruise/pkg/util/fieldindex"
+)
+
+var (
+	concurrentReconciles        = 3
+	controllerKind              = appsv1alpha1.SchemeGroupVersion.WithKind("ImageListPullJob")
+	slowStartInitialBatchSize   = 3
+	controllerName              = "imagelistpulljob-controller"
+	resourceVersionExpectations = expectations.NewResourceVersionExpectation()
+	scaleExpectations           = expectations.NewScaleExpectations()
+)
+
+// Add creates a new ImageListPullJob Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	if !utildiscovery.DiscoverGVK(controllerKind) || !utilfeature.DefaultFeatureGate.Enabled(features.KruiseDaemon) {
+		return nil
+	}
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) *ReconcileImageListPullJob {
+	return &ReconcileImageListPullJob{
+		Client:   utilclient.NewClientFromManager(mgr, controllerName),
+		scheme:   mgr.GetScheme(),
+		clock:    clock.RealClock{},
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+}
+
+// add a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r *ReconcileImageListPullJob) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r, MaxConcurrentReconciles: concurrentReconciles})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to ImageListPullJob
+	err = c.Watch(&source.Kind{Type: &appsv1alpha1.ImageListPullJob{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			oldObj := e.ObjectOld.(*appsv1alpha1.ImageListPullJob)
+			newObj := e.ObjectNew.(*appsv1alpha1.ImageListPullJob)
+			return newObj.DeletionTimestamp.IsZero() && oldObj.Generation != newObj.Generation
+		},
+	})
+	if err != nil {
+		return err
+	}
+	// Watch for changes to ImagePullJob
+	err = c.Watch(&source.Kind{Type: &appsv1alpha1.ImagePullJob{}}, &imagePullJobEventHandler{
+		enqueueHandler: handler.EnqueueRequestForOwner{
```

Comment:
it is not necessary to enqueue pull job, if the pull job status does not change significantly. For example the resulting imagelistpulljob will not change if  number of failed nodeimage changes from 1 to 2 
2. we can add a TODO here